packages/@aws-cdk/core/lib/arn.ts

```diff
@@ -268,3 +284,45 @@ function parseToken(arnToken: string, sep: string = '/', hasName: boolean = true
     return { partition, service, region, account, resource, resourceName, sep };
   }
 }
+
+
+/**
+ * Validate that a string is either unparseable or looks mostly like an ARN
+ */
+function parseArnShape(arn: string): 'unparseable' | string[] {
+  const components = arn.split(':') as Array<string>;
+  const looksLikeArn = arn.startsWith('arn:') && components.length >= 6 && components.length <= 7;
+  if (Token.isUnresolved(arn) && !looksLikeArn) {
+    return 'unparseable';
```

Comment:
why not just `undefined`?

Reply:
Why yes `undefined`?

How is this not more clear?

Reply:
Why do something different from what we usually do? 
Returning `undefined` or throwing an error is the standard practice. What's the value of deviation?

Reply:
Returning `undefined` is something we do because more expressive types are not supported in jsii. Since this piece of code is not hamstrung by those constraints, I don't see the point.

The value of deviation is that the return value is trying to connotate something other than what  `undefined` typically means. Generally, `undefined` means some kind of failure condition. That's not what I'm trying to convey here, though. This is a valid situation that needs to be handled by the calling code, just in a different way than if the ARN was parseable. 

Granted, the code word I chose was not very representative of that sentiment, because it sure looks like it represents an error. I will change it to `'token'` to represent that case better.

Alternatively, I can go full sum types and do something like:

```ts
{ shape: 'token' } | { shape: 'components'; components: string[] }
```

I had just though to short-circuit the above into:

```ts
'token' | string[]
```

Without much loss of generality or readability.